pkg/fleet-manager/fleet_plugin.go

```diff
@@ -0,0 +1,492 @@
+		}
+		if !capiutil.HasOwnerRef(secret.OwnerReferences, stsOwnerReference) {
+			secret.OwnerReferences = append(secret.OwnerReferences, stsOwnerReference)
+			if _, err := kclient.KubeClient().CoreV1().Secrets(MonitoringNamespace).Update(ctx, secret, metav1.UpdateOptions{}); err != nil {
+				return err
+			}
+		}
+	}
+
+	thanosHelmRelease := &hrapiv2b1.HelmRelease{}
+	thanosHelmReleaseNN := types.NamespacedName{Namespace: fleet.Namespace, Name: "thanos"}
+	if err := f.Client.Get(context.Background(), thanosHelmReleaseNN, thanosHelmRelease); err != nil {
+		log.Error(err, "failed to get thanos helm release")
+		return err
+	}
+
+	ownerReference := metav1.OwnerReference{
+		APIVersion: "helm.toolkit.fluxcd.io/v2beta1",
+		Kind:       hrapiv2b1.HelmReleaseKind,
+		Name:       thanosHelmRelease.Name,
+		UID:        thanosHelmRelease.UID,
+	}
+
+	svc := &corev1.Service{}
+	svcNN := types.NamespacedName{Namespace: fleet.Namespace, Name: "thanos-sidecar-remote"}
+	if err := f.Client.Get(context.Background(), svcNN, svc); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+
+		svc = &corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      svcNN.Name,
+				Namespace: svcNN.Namespace,
+				Labels:    fleetMetricResourceLables(fleet.Name),
+				OwnerReferences: []metav1.OwnerReference{
+					ownerReference,
+				},
+			},
+			Spec: corev1.ServiceSpec{
+				Ports: []corev1.ServicePort{
+					{
+						Name:        "grpc",
+						Port:        10901,
+						Protocol:    corev1.ProtocolTCP,
+						AppProtocol: pointer.String("grpc"),
+					},
+				},
+				ClusterIP: NoneClusterIP,
+			},
+		}
+
+		if err := f.Client.Create(context.Background(), svc); err != nil {
+			return err
+		}
+	}
+
+	if !capiutil.HasOwnerRef(svc.OwnerReferences, ownerReference) {
+		svc.OwnerReferences = append(svc.OwnerReferences, ownerReference)
+		if err := f.Client.Update(context.Background(), svc); err != nil {
+			return err
+		}
+	}
+
+	ep := &corev1.Endpoints{}
+	if err := f.Client.Get(context.Background(), svcNN, ep); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+
+		ep = &corev1.Endpoints{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      svcNN.Name,
+				Namespace: svcNN.Namespace,
+			},
+			Subsets: convertToSubset(endpoints),
+		}
+
+		if err := f.Client.Create(context.Background(), ep); err != nil {
+			return err
+		}
+	}
+
+	subsets := convertToSubset(endpoints)
+	if reflect.DeepEqual(ep.Subsets, subsets) {
+		return nil
+	}
+
+	ep.Subsets = subsets
+	if err := f.Client.Update(context.Background(), ep); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (f *FleetManager) reconcileObjstoreSecret(ctx context.Context, kclient *kclient.Client, secret *corev1.Secret) error {
+	_, err := kclient.KubeClient().CoreV1().Namespaces().Get(ctx, secret.Namespace, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		_, err := kclient.KubeClient().CoreV1().Namespaces().Create(ctx, &corev1.Namespace{
```

Comment:
is namespace not created by the resource?

Reply:
At this time(before HelmRelease created), you cannot assume namespace exists.